ambry-router/src/main/java/com/github/ambry/router/CompressionService.java

```diff
@@ -250,32 +248,14 @@ public ByteBuf compressChunk(ByteBuf chunkBuffer, boolean isFullChunk) {
       return null;
     }
 
-    // Convert ByteBuf to byte[] before applying compression.
-    byte[] sourceBuffer;
-    int sourceBufferOffset;
-    try {
-      Triple<byte[], Integer, Integer> bufferInfo = convertByteBufToByteArray(chunkBuffer);
-      sourceBuffer = bufferInfo.getLeft();
-      sourceBufferOffset = bufferInfo.getMiddle();
-      sourceDataSize = bufferInfo.getRight();
-    } catch (Exception ex) {
-      logger.error(String.format("Cannot compress source data because failed to convert ByteBuf to byte[]."
-              + "chunkBuffer.hasArray() = %s, chunkBuffer.readableBytes = %d, chunkBuffer.arrayOffset = %d",
-          chunkBuffer.hasArray(), chunkBuffer.readableBytes(), chunkBuffer.arrayOffset()), ex);
-
-      // Emit metrics for count how many compressions skipped due to buffer conversion error.
-      compressionMetrics.compressErrorRate.mark();
-      compressionMetrics.compressErrorBufferConversion.inc();
-      return null;
-    }
-
     // Apply compression.
     CompressionMetrics.AlgorithmMetrics algorithmMetrics = compressionMetrics.getAlgorithmMetrics(defaultCompressor.getAlgorithmName());
-    Pair<Integer, byte[]> compressResult;
+    ByteBuf newChunkBuffer = combineBuffer(chunkBuffer);
```

Comment:
we don't have to return a ByteBuf from this method, we can just return a ByteBuffer. so the combineBuffer can be
```
    // If the source buffer is already has 1 buffer, create a duplicate.
    if (byteBuf.nioBufferCount() == 1) {
      return byteBuf.nioBuffer();
    }

    // Copy the buffer out without changing index.
    ByteBuffer newBuffer = ByteBuffer.allocateDirect(byteBuf.readableBytes());
    byteBuf.readBytes(newBuffer);
    newBuffer.flip();
    return newBuffer;
```